cmstestsuite/unit_tests/grading/__init__.py

```diff
@@ -0,0 +1,174 @@
+#!/usr/bin/env python2
+# -*- coding: utf-8 -*-
+
+# Contest Management System - http://cms-dev.github.io/
+# Copyright © 2018 Stefano Maggiolo <s.maggiolo@gmail.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Tests for common grading functions.
+
+"""
+
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import StringIO
+import unittest
+
+from cms.grading import Sandbox, WHITES, \
+    format_status_text, merge_evaluation_results, white_diff
+
+
+class TestFormatStatusText(unittest.TestCase):
+
+    @staticmethod
+    def _tr(s):
+        return s.replace("A", "E")
+
+    def test_success_no_placeholders(self):
+        self.assertEqual(format_status_text([""]), "")
+        self.assertEqual(format_status_text(["ASD"]), "ASD")
+        self.assertEqual(format_status_text(["你好"]), "你好")
+
+    def test_success_with_placeholders(self):
+        self.assertEqual(format_status_text(["%s", "ASD"]), "ASD")
+        self.assertEqual(format_status_text(["ASD%s\n%s", "QWE", "123"]),
+                         "ASDQWE\n123")
+
+    def test_success_with_translator(self):
+        # Translation is applied before formatting.
+        self.assertEqual(format_status_text(["A%s", "ASD"], self._tr), "EASD")
+        self.assertEqual(
+            format_status_text(["AAA %s\n%s", "AAA", "AE"], self._tr),
+            "EEE AAA\nAE")
+
+    def test_insuccess(self):
+        self.assertEqual(format_status_text(["%s"]), "N/A")
+        self.assertEqual(format_status_text(["%s"], self._tr), "N/E")
+
+
+class TestMergeEvaluationResults(unittest.TestCase):
+
+    @staticmethod
+    def _res(execution_time, execution_wall_clock_time, execution_memory,
+             exit_status, signal=None, filename=None):
+        r = {
+            "execution_time": execution_time,
+            "execution_wall_clock_time": execution_wall_clock_time,
+            "execution_memory": execution_memory,
+            "exit_status": exit_status,
+        }
+        if signal is not None:
+            r["signal"] = signal
+        if filename is not None:
+            r["filename"] = filename
+        return r
+
+    def test_success_status_ok(self):
+        self.assertEqual(
```

Comment:
Isn't it bad practice to compare floats for exact equality? Looking at the doc I see that starting with Python 3.1 special type-specific comparators are used. So maybe here we use a dict comparator, that then in turn uses a float approximate comparator? Could you check?

Reply:
Done. The dict-specific is the same as assertEqual but in addition tests they are dicts, so no special float comparison.